### ISSUE TITLE
bert-for-tf2 for Sentiment Analysis

### ISSUE THREAD

**Issue by muhammadfahid51** [kpe/bert-for-tf2#33]
Hi,
Can anyone provide me with a guide on how to use bert-for-tf2 for a custom task like sentiment analysis. 
I have been trying but to no avail. My model is training but its not giving me any accuracy. 

Below is the method I am using

```
model_name = "multi_cased_L-12_H-768_A-12"
model_dir = "models/multi_cased_L-12_H-768_A-12"
model_ckpt = os.path.join(model_dir, "bert_model.ckpt")
bert_params = bert.params_from_pretrained_ckpt(model_dir)
l_bert = bert.BertModelLayer.from_params(bert_params, name="bert", trainable=True)

l_input_ids      = tf.keras.layers.Input(shape=(max_seq_len,), dtype='int32')
l_token_type_ids = tf.keras.layers.Input(shape=(max_seq_len,), dtype='int32')

l_bert_o = l_bert(l_input_ids)
conv_blocks = []
for k_size in FILTER_SIZES:
    conv = tf.keras.layers.Conv1D(filters=NUM_FILTERS,
                         kernel_size=k_size,
                         padding="valid",
                         activation="relu",
                         strides=1)(l_bert_o)
    conv = tf.keras.layers.MaxPooling1D(pool_size=max_seq_len - k_size + 1)(conv)
    conv = tf.keras.layers.Flatten()(conv)
    conv_blocks.append(conv)
concat = tf.keras.layers.Concatenate()(conv_blocks) if len(conv_blocks) > 1 else conv_blocks[0]
concat = tf.keras.layers.Dropout(0.2)(concat)
x = tf.keras.layers.Dense(256, activation="relu")(concat)
output = tf.keras.layers.Dense(3, activation="softmax")(x)
model = tf.keras.models.Model(inputs=l_input_ids, outputs=output)

model.compile(loss="categorical_crossentropy", optimizer="rmsprop",
              metrics=["accuracy"])

bert.load_bert_weights(l_bert, model_ckpt)
model.summary()

```

**Comment by kpe** [comment]
@muhammadfahid51 - have you took a look at the examples, i.e. https://github.com/kpe/bert-for-tf2/blob/master/examples/tpu_movie_reviews.ipynb

I'm not sure whether doing something fancy on top of BERT when fine-tuning brings much (compared to just classifying on the [CLS] activations). I believe it is rather important to try not over-fitting and find a good learning rate schedule.

**Comment by kpe** [comment]
@muhammadfahid51 - what do you mean by "model is training" - does the loss drop over time? What hyper parameters do you use? What is your kernel size in the CNN and how many CNN layers do you use? And how large is your dataset?
I don't see why your model above should not work if using proper hyper-parameters.

**Comment by muhammadfahid51** [comment]
@kpe Hey, 
Training loss does not drop over time and neither the validation loss. About parameters I have changed them alot. Like tried different number of layers, filters, regularization parameters etc but to no avail. 
But I think the problem was something else. My data is in Urdu Language and it has got 10k samples of 3 classes with 50(Positive), 30(Negative) and 20(Neutral) percent distribution. I tried the same model on imdb data and it worked well. 
But one tweak that remains is,
How does bert handles unknown words. Does it assign all of them to '[UNK]' and then learns them while training or is there some other technique ?

**Comment by kpe** [comment]
@muhammadfahid51 - I believe the BERT models were pre-trained on English only. So until there is a multilingual pre-trained model you might have to pre-train from scratch yourself. To check how the tokenization works with your language, you could tokenize and convert the tokens to ids, and then back. 
Currently bert-for-tf2 does not have an example/tools for pre-training from scratch. When pre-training from scratch it might be easier using the workflow from the original google-research/bert. Once you have the pre-trained weights, you could fine-tune with bert-for-tf2.
